### ISSUE TITLE
[DRY] why is `_≉_` defined both for `Algebra.Bundles.Raw.RawX` bundles, and via `Setoid` instances, for `Algebra.Bundles.X`?

### ISSUE THREAD

**Issue by jamesmckinna** [agda/agda-stdlib#2274]
This is another instance of the issues arising when adding 'derived forms' to a(n algebraic) structure/bundle, namely where should they be added:
* to the `RawX` bundle? this is currently the case for `_≉_`, but AFAICT, never used, except to hereditarily be re-exported to richer structures;
* to the `IsX` structure? *not* currently the case for  `_≉_` via an `IsEquivalence` instance, but is for `IsGroup._-_` cf. #2251 ;
* to the `X` bundle? this is *also* the case for `_≉_` via the `Setoid` instance (#2095 , #2162 ), which is in fact a 'derived form' defined in the `IsX` structure... *and* by the opening of the underlying `RawX` bundle.

The first and third cases seem to me to represent a clear DRY violation, and probably arise from the accumulation of  'legacy' design decisions. But what, if any, should be our refactoring policy towards a  revised design?

**Comment by jamesmckinna** [comment]
In the absence of further discussion yet, note there are two possible scenarios:
* where the derived operation indeed is derivable from the `Raw` *signature* (these are the 'easy' examples being considered under #2251 / #2268 );
* where the operation is only definable *under the assumptions that the `IsX` axiomatisation holds* (for some reason, the first example that I can think of is the 'snake lemma' for Abelian categories... but presumably 'easier' naturally occurring examples arise for theories simpler than the essentially-algebraic fragment? Maybe not!?)

I'm not sure what the official (universal algebra) nomenclature is for each kind, but on the model of the theory of formal systems, I supposed the first should be called 'derivable', while the second 'admissible'?

In any case, this issue concerns derivable operations being defined 'too late' to be exported to any subsequently definable extension of the algebra. And that does seem to be a DRY problem... over and above the 'where should such things live?' question...

**Comment by JacquesCarette** [comment]
Finally catching up on really old stuff - my mailbox is quite full of saved email about items that I meant to comment on.  Still have 30 or so from 2024 alone, never mind a whole bunch from 2023!

I do agree that the library is over-eager to define 'helper' conservative extensions that turn out to be seldom used. (I don't know what the official nomenclature is either, but 'derivable' and 'admissible' make a lot of sense.)

I'm a bit concerned about solving DRY problems by over-defining things just on the off chance they might be needed. (Especially here as negation is so so often a code smell in constructive mathematics.)

Personally I'd prefer some 'helper' module(s) [with a better name than Helper !] that define extensions in a standard way. I've migrate things towards that in `agda-categories` -- it has really helped reduce bloat. (`Monoidal` was criminally bloated, which had non-trivial downstream effects.)

**Comment by jamesmckinna** [comment]
@JacquesCarette thanks for returning to this! And I appreciate you separating out two parts to this:
* what notions should be added automatically?
* where should they be added?

But here the problem is that, independent of any answer to the first, the second is answered in weird/bad/DRY-violating ways.

As to the first, I'm a bit 'wider' than you as regards extending `record` APIs, here out of a consideration for *consistency of notation*, as well as relieving the user of the burden of *ad hoc* invention, esp. in the context of a 'standard' library (along the lines of 'batteries included': 'if you `open` this, then you'll get all of this extra, for free'), and I'm willing to take the space hit... I think "bloat" is too strong here!

So a resolution of this issue for me would happily resolve the first question in whatever form (we should reach agreement... either as an 'extension' module, or as 'batteries included', and if so, how much?), but first  how we consider appropriate to resolve/standardise the second...

**Comment by jamesmckinna** [comment]
As for `_≉_` itself, it's perhaps unfortunate to have it in the title of the issue, which partly reflects that the underlying `Setoid` structures of an `Algebra` structure/bundle are defined under `Relation.Binary`, and, moreover, are not defined relative to a `Raw` (equality) relation (single field records tend not to get  defined in the library, except for `instance` inference, and the `Unit` type as a special case...?), so it seems as though there isn't an obvious place to add negated equality as an additional manifest field... 

... but your point about it being a code smell is an interesting one. Barring eg it's 'general' properties wrt `Respects` mentioned in #281

**Comment by JacquesCarette** [comment]
Through the many many words, talking about many fascinating issues... I've lost the thread about what is the main issue at stake here?

**Comment by MatthewDaggitt** [comment]
My view is that really the correct solution is number 1. It should be in the `Raw` bundles. I think the missing part of the puzzle is that we don't have a `RawSetoid` bundle in which to put the negated equality operator...

**Comment by jamesmckinna** [comment]
> Through the many many words, talking about many fascinating issues... I've lost the thread about what is the main issue at stake here?

Yup, that's my weakness... (too) "many many words" ;-)

TL;DR: there are *2* issues at stake:
* (general) where should admissible/derivable operations live, so that we correctly export them in sub-structures/bundles; in general, I believe that derivable operations should live in the `Raw` bundle (but `IsGroup` is an anomaly, so `_//_`, `_\\_` and `_-_` should be moved to the `RawGroup` bundle...)
* (particular) the derivable `_≉_` is an anomaly, for the reasons that @MatthewDaggitt points out (because `RawSetoid` doesn't exist; another possible downstream PR? in which case where should it live: under `Relation.Binary.Bundles.Raw` which would be new, or under `Algebra.Bundles.Raw`)

In each case, the solution would (seem to) be a fiddly (and possibly, for the `IsGroup` case, `breaking`) PR moving things around/adding new things, so suggest tagging this as v3.0 when we might reconsider this and eg. #2252 #2255 #2268 (though this could be dealt with independently in v2.1 as a prototype of how to fix it in general...?)

**Comment by Taneb** [comment]
`RawSetoid`, if we want to add it, should live in `Relation.Binary.Bundles.Raw`, and it should probably have some friends in there with it

**Comment by jamesmckinna** [comment]
@Taneb yup, all the various ordering relations and their flipped- and negated- versions should be in `Raw` bundles for such things... so maybe a v2.1 version is possible/desirable?

**Comment by jamesmckinna** [comment]
Looping back to @JacquesCarette 's comment above, in the light of #2491 

> I'm a bit concerned about solving DRY problems by over-defining things just on the off chance they might be needed. (Especially here as negation is so so often a code smell in constructive mathematics.)
> 
> Personally I'd prefer some 'helper' module(s) [with a better name than Helper !] that define extensions in a standard way. I've migrate things towards that in `agda-categories` -- it has really helped reduce bloat. (`Monoidal` was criminally bloated, which had non-trivial downstream effects.)

I think that perhaps, along the 'minimalist'/'maximalist' spectrum of `library-design`, that you *might* be right in leaning towards 'minimalist'... *as regards properties*. Here, I'm more concerned with *vocabulary* ... which feels 'lighter' (as regards 'bloat'), while at the same time providing clients with 'just enough' extra structure that they don't end up reinventing the wheel?

**Comment by jamesmckinna** [comment]
While working on #2491 I had hoped that `Relation.Binary.Properties.ApartnessRelation` could be refactored to make better use of the negated apartness symbol defined in the `Raw` bundle... but in fact it needs to be defined in the `IsApartnessRelation` structure in order to be used in `Relation.Binary.Properties.ApartnessRelation`.

So @JacquesCarette 's general point about 'helper' modules actually seems to suggest that we should prefer option 2 in the original proposal above, because the 'helper module' associated with an `IsX` structure (in `*.Properties.X`) does *not* depend on the extended `Raw` bundle... and trying to retrofit that would seem very strange to me... unless we go for the `hypothetical-rewrite` proposal in #2252 ...!?

**Comment by JacquesCarette** [comment]
Depending on what you mean by _vocabulary_, I've moved some of that out too -- see the sub-module `ShortHands` in [Categories.Category.Monoidal.Utilities](https://agda.github.io/agda-categories/Categories.Category.Monoidal.Utilities.html)!

I used to think that defining all sorts of helper stuff early and 'globally' was a great idea, as I thought that this had low 'weight'. Turns out this is not true in current Agda (but could be a deficiency?) So I've moved to pulling these out, at least in the worst offenders.

**Comment by jamesmckinna** [comment]
Thanks @JacquesCarette I think that Matthew has commented elsewhere as to the excess 'weight' of adding *properties* to `IsX` records. Interesting that even introducing symbols/derived forms might be undesirable. Of course, we could agree not to merge any PR moving towards the proposal here until this discussion has bottomed-out.

**Comment by JacquesCarette** [comment]
Coming back to this: adding some small amount of very standard vocabulary might make sense. We'd have to have guidelines. And yes, vocabulary is different than properties.

I was just shocked by how much "standard vocabulary" came with a Monoidal Category -- and how much that bloated things downstream.

**Comment by jamesmckinna** [comment]
Not sure if #2491 closes this, or whether it should be kept open for the above more general discussion, and pending (resolution of) the `IsEquivalence` vs. `Setoid` discussion in #2499 (ie whether `Algebra.Structures` should export `(Raw)Setoid`s at all...)?

**Comment by JacquesCarette** [comment]
Unfortunately I get the feeling that the issue outlined here isn't quite all resolved (yet). If nothing else, there's some design to document.

**Comment by jamesmckinna** [comment]
Thanks @JacquesCarette .

I think that there are two things to do to resolve the [ DRY ] aspect, and separately, to leave this issue open regarding the wider question of the minimal/maximal 'basic' API for `Algebra`:
- [x] adding `RawSetoid` etc. and hence moving the question of such API design choices to (potential refactorings of) `Algebra.Bundles.Raw`; I think #2491 does achieve this
- [ ] adding a `rawSetoid` manifest field to `Algebra.Structures.*`, and, via `open public`, bringing the existing negated equality symbol into scope, and leaving open what, if anything, might additionally be brought into scope in a downstream refactoring

Regarding the second, I think this can be done with a minimally invasive PR, in preparation now.

**Comment by JacquesCarette** [comment]
Yes, I do think that that PR does the job. See additional comments there.

**Comment by jamesmckinna** [comment]
Now that this is linked to from #3053 , and possibly/probably closed by #2536 , closing this now in this interests of tidying up the `Issues` queue/stack...
